src/main/java/core/stack/MateStack.java

```diff
@@ -1,19 +1,44 @@
 package core.stack;
 
+import java.util.EmptyStackException;
+
 public class MateStack<T> {
-    public void push(T value) {
+    private static class Node<T> {
+        private final T value;
+        private Node<T> prev;
+
+        private Node(T value) {
+            this.value = value;
+        }
+    }
+
+    private Node<T> top = null;
+    private int size = 0;
 
+    public void push(T value) {
+        Node<T> top = new Node<>(value);
+        top.prev = this.top;
+        this.top = top;
+        size++;
```

Comment:
I would suggest creating and using constructor with all fields, something like this:

```suggestion
        Node<T> first = head;
        head = new Node<>(first, value);
        size++;
```